### ISSUE TITLE
`dist` seems to ignore environmnet variables configured in `.cargo/config.toml`

### ISSUE THREAD

**Issue by jan-ferdinand** [axodotdev/cargo-dist#1571]
tl;dr: [minimally failing example](https://github.com/jan-ferdinand/dist-this)

### Description

In a project I'm working on, we're using [`tokio`'s unstable features](https://docs.rs/tokio/1.41.1/tokio/index.html#unstable-features). Those features require opt-in through an environmnent variable. This environment variable can be set in the file `.cargo/config.toml`, which is what we do. Unfortunately, `dist build` seems to ignore this configuration. As a result, the compilation step during `dist build` fails.

### What I've tried

- Running `RUSTFLAGS="--cfg tokio_unstable" dist build`[^1] works fine.
- I cannot [find](https://opensource.axo.dev/cargo-dist/book/?search=environment%20variable) anything applicable to this situation in the `dist` book.
- I was unable to identify similar issues here on github.

### Questions

1. Do you consider this a bug or missing feature, or am I miss- or even abusing `dist`?
2. Is there a maintainable workaround? I feel uncomfortable monkey-patching the `release.yml` to jam this environment variable in there every time I run `dist init`.
3. Do you need or want any additional information?

[^1]: as you might expect, `RUSTFLAGS="--cfg tokio_unstable"` is the environment variable configured in `.cargo/config.toml`

**Comment by ashleygwilliams** [comment]
hello @jan-ferdinand ! thank you so much for the detailed issue *and* repro, it's super appreciated. offhand, this feels like a bug to me as i know we have code to handle RUSTFLAGS (https://github.com/axodotdev/cargo-dist/blob/main/cargo-dist/src/build/cargo.rs#L73).

i agree that editing the `release.yml` file is very undesirable. if in the meantime you need to, we do offer the `allow-dirty` flag but, this is something to use as a last resort.

you *may* be able to use the [`custom-build-step` feature](https://opensource.axo.dev/cargo-dist/book/ci/customizing.html#customizing-build-setup) to work around this for now, though i think we'd want to fix this as i think it's a bug.

out of curiousity- is the project you are building part of a workspace? by default, we run `cargo build --workspace` and i wonder if maybe that is why the cargo config is getting dropped? we have a feature called `precise-builds` that you can use to change that build command to build each package individually (this has some potentially negative edge case side effects, but it would be a good experiment).

**Comment by jan-ferdinand** [comment]
Thanks for your speedy reply & help!

> offhand, this feels like a bug to me as i know we have code to handle RUSTFLAGS

The `FIXME:` comment does indeed suggest that dealing with `RUSTFLAGS` is both [complex](https://doc.rust-lang.org/cargo/reference/config.html) and not yet complete. Perhaps a library like [`cargo-config2`](https://github.com/taiki-e/cargo-config2) could help? I have no experience with it, but its description and the example in the readme sound like it could be a good fit.

> you _may_ be able to use the [custom-build-step feature](https://opensource.axo.dev/cargo-dist/book/ci/customizing.html#customizing-build-setup) to work around this for now

Thanks for the pointer. I will play around and see if I can get something going.

> is the project you are building part of a workspace?

The project where I stumbled across this is a workspace, yes. The minimal failing example repository is not. This seems to suggest that workspaces having nothing to do with it, but this is not a very educated guess. Of course, different bugs could cause seemingly identical behavior.

**Comment by jan-ferdinand** [comment]
I have managed to [work around](https://github.com/jan-ferdinand/dist-this/commit/c19bef294f29846a9ab2331a5f75c0ec4798df31) the concrete issue using a custom build step feature, as thankfully suggested by @ashleygwilliams. Should somebody else stumble across this issue between now and a fix for the underlying bug, here are some gottchas I ran into:[^experiment]
- As far as I can see, the custom build step workflow file cannot be syntactical, introducing an always-failing workflow run.
- The location for the `github-build-setup` key is correctly documented [here](https://opensource.axo.dev/cargo-dist/book/reference/config.html#github-build-setup), but incorrectly documented [here](https://opensource.axo.dev/cargo-dist/book/ci/customizing.html#customizing-build-setup).
- Things break if too many (nested & escaped) quotation marks are present in the custom build step.
```yaml
# this works
  run: echo RUSTFLAGS="$RUSTFLAGS --cfg tokio_unstable" >> "$GITHUB_ENV"
# this does not work
  run: echo "RUSTFLAGS=\"$RUSTFLAGS --cfg tokio_unstable\"" >> "$GITHUB_ENV"
```
The version that doesn't work is closer to the [documented](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/workflow-commands-for-github-actions#setting-an-environment-variable) way of setting an environment variable in a github workflow.

[^experiment]: Of course, the custom build step feature _is_ documented as experimental.

**Comment by RagnarGrootKoerkamp** [comment]
Created #2202 to possibly fix this, by using `cargo_config2` as suggested.

Indeed, I'm also in the situation where I would like to preserve/propagate `rustflags` from my `.config/cargo.toml`, specifically because it depends on the target, and so exporting that to a target-dependent environment variable is annoying.

**Comment by somtochiama** [comment]
Just bumping this. I also have a project with different rustflags for different targets, and the env variable override is making it difficult to use cargo-dist.
